Quantum-Control-Applications/Superconducting/Two-Flux-Tunable-Coupled-Transmons/Use Case 4 - Single- and Two-Qubit State and Process Tomography/two-qubit-process-tomography.py

```diff
@@ -0,0 +1,331 @@
+            wait(pi_len // 4, f"q{qubit2}_xy")
+        with case_(1):
+            play("x180", f"q{qubit2}_xy")
+        with case_(2):
+            play("y90", f"q{qubit2}_xy")
+        with case_(3):
+            play("-y90", f"q{qubit2}_xy")
+        with case_(4):
+            play("-x90", f"q{qubit2}_xy")
+        with case_(5):
+            play("x90", f"q{qubit2}_xy")
+
+
+# subroutine to switch between the tomographically-complete
+# set of bases in which to measure
+def measurement_basis_change(k, l, qubit1, qubit2):
+
+    # Change basis with operation k on qubit1
+    with switch_(k):
+        with case_(0):
+            wait(pi_len // 4, f"q{qubit2}_xy")
+        with case_(1):
+            play("x180", f"q{qubit1}_xy")
+        with case_(2):
+            play("y90", f"q{qubit1}_xy")
+        with case_(3):
+            play("-y90", f"q{qubit1}_xy")
+        with case_(4):
+            play("-x90", f"q{qubit1}_xy")
+        with case_(5):
+            play("x90", f"q{qubit1}_xy")
+
+    # Change basis with operation l on qubit2
+    with switch_(l):
+        with case_(0):
+            wait(pi_len // 4, f"q{qubit2}_xy")
+        with case_(1):
+            play("x180", f"q{qubit2}_xy")
+        with case_(2):
+            play("y90", f"q{qubit2}_xy")
+        with case_(3):
+            play("-y90", f"q{qubit2}_xy")
+        with case_(4):
+            play("-x90", f"q{qubit2}_xy")
+        with case_(5):
+            play("x90", f"q{qubit2}_xy")
+
+
+with program() as two_qubit_process_tomography:
+
+    n = declare(int)
+    n_st = declare_stream()
+
+    I1 = declare(fixed)
+    Q1 = declare(fixed)
+    I2 = declare(fixed)
+    Q2 = declare(fixed)
+
+    state = declare(bool)  # QUA variable for the measured qubits state
+    state_st = declare_stream()  # Stream for the qubits state
+
+    c1 = declare(int)  # QUA variable for switching between state preparation/creations on qubit 1
+    c2 = declare(int)  # QUA variable for switching between state preparation/creations on qubit 2
+    m1 = declare(int)  # QUA variable for switching between Bloch basis projections/measurements on qubit 1
+    m2 = declare(int)  # QUA variable for switching between Bloch basis projections/measurements on qubit 2
+
+    with for_(n, 0, n < n_avg, n + 1):  # QUA for_ loop for averaging
+        with for_(c1, 0, c1 <= 5, c1 + 1):  # QUA for_ loop for switching between state preparations on qubit 1
+            with for_(c2, 0, c2 <= 5, c2 + 1):  # QUA for_ loop for switching between state preparations on qubit 2
+                with for_(
+                    m1, 0, m1 <= 5, m1 + 1
+                ):  # QUA for_ loop for switching between Bloch basis projections/measurements on qubit 1
+                    with for_(
+                        m2, 0, m2 <= 5, m2 + 1
+                    ):  # QUA for_ loop for switching between Bloch basis projections/measurements on qubit 2
+
+                        reset_frame(f"q{qubit1}_xy")
```

Comment:
same as in previous programs, for now the process does not involve frame rotations, so no need to reset_frame. The same goes for reset_phase, as in the two-qubit-state-tomography, so we can let these four lines of code go :)

Reply:
@KevinAVR, are we sure this is true for the two-qubit case? We found that the chi matrix plots were incorrect when performing two-qubit process tomography for an iSWAP gate, when we didn't include the phase and frame resets, but were correct after resetting the phase and frame . . .

Reply:
Hi @bornman-nick I did not think of the two qubit gate case, since I only saw examples of single qubit gates. Yes indeed a two qubit gate can bring more intricasies so lets keep as you suggested :) I don't want to block further the review :)

Reply:
Yeah, the state preparation and measurement basis change steps use only single-qubit gates, but the process being investigated could obviously involve some entangling operation. Thanks for checking this all!